io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -1079,130 +1090,290 @@ traceMany ((time, tid, event):ts) trace =
-      ReturnStm x     -> do (vids, tids) <- finaliseCommit written
-                            return (StmTxComitted x vids tids nextVid)
-      ThrowStm e      -> do finaliseAbort written
-                            return (StmTxAborted (toException e))
-      Retry           -> do vids <- finaliseRetry read written
-                            return (StmTxBlocked vids)
-      NewTVar x k     -> do v <- execNewTVar nextVid x
-                            go read written (succ nextVid) (k v)
-      ReadTVar v k    -> do x <- execReadTVar v
-                            go (SomeTVar v : read) written nextVid (k x)
-      WriteTVar v x k -> do execWriteTVar v x
-                            go read (SomeTVar v : written) nextVid k
-
-    -- Revert all the TVar writes
-    finaliseAbort written =
-      sequence_ [ revertTVar  tvar | SomeTVar tvar <- reverse written ]
-
-    -- Revert all the TVar writes and put this thread on the blocked queue for
-    -- all the TVars read in this transaction
-    finaliseRetry :: [SomeTVar s] -> [SomeTVar s] -> ST s [TVarId]
-    finaliseRetry read written = do
-      sequence_ [ revertTVar  tvar | SomeTVar tvar <- reverse written ]
-      sequence_ [ blockOnTVar tvar | SomeTVar tvar <- read ]
-      return [ vid | SomeTVar (TVar vid _ _ _) <- read ]
-
-    blockOnTVar :: TVar s a -> ST s ()
-    blockOnTVar (TVar _vid _vcur _vsaved blocked) =
-      --TODO: avoid duplicates!
-      modifySTRef blocked (mytid:)
-
-
-execAtomically' :: StmA s () -> ST s [(ThreadId, [TVarId])]
-execAtomically' = go []
+    go :: forall b.
+          StmStack s b a
+       -> Map TVarId (SomeTVar s)  -- set of vars read
+       -> Map TVarId (SomeTVar s)  -- set of vars written
+       -> [SomeTVar s]             -- vars written in order (no dups)
+       -> TVarId                   -- var fresh name supply
+       -> StmA s b
+       -> ST s (StmTxResult s a)
+    go ctl !read !written writtenSeq !nextVid action = assert localInvariant $
+                                                       case action of
+      ReturnStm x -> case ctl of
+        AtomicallyFrame -> do
+          -- Commit each TVar
+          traverse_ (\(SomeTVar tvar) -> do
+                        commitTVar tvar
+                        -- Also assert the data invariant that outside a tx
+                        -- the undo stack is empty:
+                        undos <- readTVarUndos tvar
+                        assert (null undos) $ return ()
+                    ) written
+
+          -- Return the vars written, so readers can be unblocked
+          return (StmTxCommitted x (reverse writtenSeq) nextVid)
+
+        OrElseLeftFrame _b k writtenOuter writtenOuterSeq ctl' -> do
+          -- Commit the TVars written in this sub-transaction that are also
+          -- in the written set of the outer transaction
+          traverse_ (\(SomeTVar tvar) -> commitTVar tvar)
+                    (Map.intersection written writtenOuter)
```

Comment:
it is not clear why we commit here only the intersection and not all `written`.  I checked that either pass the tests, so maybe it's not really important.